File: src/main/java/micronaut/playground/domain/service/GreeterImpl.java
package micronaut.playground.domain.service;

import io.reactivex.Maybe;
import io.reactivex.Single;
import micronaut.playground.client.model.Hello;
import micronaut.playground.config.TemplateConfiguration;
import micronaut.playground.repository.VisitRepository;

import javax.inject.Singleton;
import javax.validation.constraints.NotEmpty;
import java.time.LocalDateTime;

@Singleton
public class GreeterImpl implements Greeter {

  private TemplateConfiguration configuration;
  private VisitRepository visitRepository;

  public GreeterImpl(TemplateConfiguration configuration, VisitRepository visitRepository) {
    this.configuration = configuration;
    this.visitRepository = visitRepository;
  }

  @Override
  public Single<Hello> greetHello(@NotEmpty String name) {
    Maybe<LocalDateTime> lastVisitSingle = Maybe
        .fromCallable(() -> visitRepository.lastVisit(name))
        .flatMap(v -> Maybe.just(v.orElse(null)));
    Single<Long> visitCountSingle = Single.just(name).map(visitRepository::countVisit);
    return Single.zip(visitCountSingle.materialize(), lastVisitSingle.materialize(),
        (count, lastVisit) -> new Hello(configuration.getPrefix(),
            name,
            count.getValue(),
            lastVisit.getValue()))
          .doOnSuccess(c -> visitRepository.addVisit(name));
  }
}

File: src/main/kotlin/micronaut/playground/domain/service/GreeterImpl.kt
package micronaut.playground.domain.service

import io.micronaut.context.annotation.Property
import micronaut.playground.client.model.Hello
import micronaut.playground.repository.VisitRepository
import jakarta.inject.Singleton
import micronaut.playground.domain.entity.Visit
import reactor.core.publisher.Mono
import java.time.LocalDateTime

@Singleton
open class GreeterImpl(val visitRepository: VisitRepository) : Greeter {

    @field:Property(name = "template.prefix")
    var prefix: String? = null

    override fun greetHello(name: String): Mono<Hello> {
        val lastVisitSingle = visitRepository.lastVisit(name)
            .defaultIfEmpty(Visit(0, name, LocalDateTime.now()))
        val visitCountSingle = visitRepository.countVisit(name)
        return Mono.zip(
            visitCountSingle, lastVisitSingle,
            { count, lastVisit -> Hello(prefix.orEmpty(), name, count, lastVisit.createdAt) })
            .doOnError { e -> e.printStackTrace() }
            .flatMap { res -> visitRepository.addVisit(Visit(name)).thenReturn(res) }
    }
}
